gridcomps/ExtData2G/ExtDataUpdatePointer.F90

```diff
@@ -36,56 +42,91 @@ subroutine create_from_parameters(this,update_time,update_freq,update_offset,tim
       class(ExtDataPointerUpdate), intent(inout) :: this
       logical, intent(out) :: do_update
-      type(ESMF_Time), intent(inout) :: working_time
+      type(ESMF_Time), intent(inout) :: use_time
       type(ESMF_Time), intent(inout) :: current_time
       logical, intent(in) :: first_time
       integer, optional, intent(out) :: rc
-      type(ESMF_Time) :: previous_ring
+      type(ESMF_Time) :: next_ring
 
       integer :: status
 
       if (this%disabled) then
          do_update = .false.
          _RETURN(_SUCCESS)
       end if
-      if (ESMF_AlarmIsCreated(this%update_alarm)) then
+      if (this%simple_alarm_created) then
+         use_time = current_time+this%offset
          if (first_time) then
-            call ESMF_AlarmGet(this%update_alarm,prevRingTime=previous_ring,__RC__)
-            working_time =previous_ring+this%offset
             do_update = .true.
+            this%first_time_updated = .true.
+            use_time = this%last_ring + this%offset
          else
-            do_update = ESMF_AlarmIsRinging(this%update_alarm,__RC__)
-            working_time = current_time+this%offset
+            ! normal flow
+            next_ring = this%last_ring
+            if (current_time > this%last_checked) then
+               do while (next_ring < current_time)
+                  next_ring=next_ring+this%update_freq
+               enddo
+               if (current_time == next_ring) then
+                  do_update = .true.
+                  this%last_ring = next_ring
+                  this%first_time_updated = .false. 
+               end if
+            ! if clock went backwards, so we must update, set ringtime to previous ring from working time
+            else if (current_time < this%last_checked) then
+               next_ring = this%last_ring
+               ! the clock must have rewound past last ring
+               if (this%last_ring > current_time) then
+                  do while(next_ring <= current_time)
+                     next_ring=next_ring-this%update_freq
+                  enddo
+                  use_time = next_ring+this%offset
+                  this%last_ring = next_ring
+               ! alarm never rang during the previous advance, only update the previous update was the first time
+               else if (this%last_ring < current_time) then
+                    if (this%first_time_updated) then
+                       do_update=.true.
+                       this%first_time_updated = .false.
+                       use_time = this%last_ring + this%offset
+                    end if
+               ! otherwise we land on a time when the alarm would ring and we would update
+               else if (this%last_ring == current_time) then
+                  do_update =.true.
+                  this%first_time_updated = .false.
+                  use_time = current_time+this%offset
+               end if
+            end if
```

Comment:
As above - I'd like to see if there are ways to simplify this.

Reply:
I have to handle multiple special cases. 
1. If this is the first time
2. If the clock moves forward and the alarm rings while moving forward, then you externally rewind the clock
3. If the clock moves forward and the alarm does not ring while moving forward, then you externally rewind the clock
4. If the clock moves forward and the alarm does not ring while moving forward, then you externally rewind the clock and this the last update was the first time running
5. Oh, and I don't trust ESMF alarms at all with the currently implementation so I have to do this without ESMF Alarms.

Reply:
Understood.     But we should still be able to introduce helper functions so that the large scale structure is simpler.  Thee key is coming up with a good name for what happens in each block.  A purist might even argue that it is worth procedures along the lines of  `this%is_now()` `this%is_past()` and `this%is_future()` or something like that.  Mostly just want it to be considered; not pushing hard.

Reply:
I'll try to introduce helper functions